src/signer/canister/src/lib.rs

```diff
@@ -164,6 +172,90 @@ pub async fn generic_sign_with_ecdsa(
 }
 
+/// Returns the Schnorr public key of the caller or specified principal.
+///
+/// Note: This is an exact dual of the canister [`schnorr_public_key`](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-schnorr_public_key) method.  The argument and response types are also the same.
+///
+/// # Arguments
+/// - `arg`: The same `SchnorrPublicKeyArgument` as the management canister argument.  The semantics
+///   are identical but the meaning of the fields in the new context deserve some explanation.
+///   - `arg.canister_id`: The principal of the canister or user for which the Chain Fusion Signer
+///     has issued the public key.  If `None`, the caller's public key is returned.
+///   - `arg.derivation_path`: The derivation path to the public key.  The caller is responsible for
+///     ensuring that the derivation path is used to namespace appropriately and to ensure that
+///     unintended sub-keys are not requested.  At minimum, it is recommended to use `vec!["NAME OF
+///     YOUR APP".into_bytes()]`.  The maximum derivation path length is 254, one less than when
+///     calling the management canister.
+///   - `arg.key_id`: The ID of the root threshold key to use.  E.g. `key_1` or `test_key_1`.  See <https://internetcomputer.org/docs/current/references/t-sigs-how-it-works#key-derivation>
+///     for details.
+/// - `payment`: The payment type to use.  If omitted or `None`, it will be assumed that cycles have
+///   been attached.
+///
+/// # Warnings
+/// - The user supplied derivation path is used as-is.  The caller is responsible for ensuring that
+///   derivation paths are used to namespace appropriately and to ensure that unintended sub-keys
+///   are not requested.
+/// - It is recommended that, at minimum, the derivation path should be `vec!["NAME OF YOUR
+///   APP".into_bytes()]`
+///
+/// # Panics
+/// - If the caller is the anonymous user.
+#[update(guard = "caller_is_not_anonymous")]
+pub async fn schnorr_public_key(
+    arg: SchnorrPublicKeyArgument,
+    payment: Option<PaymentType>,
+) -> Result<(SchnorrPublicKeyResponse,), SchnorrPublicKeyError> {
+    PAYMENT_GUARD
+        .deduct(
+            payment.unwrap_or(PaymentType::AttachedCycles),
+            SignerMethods::SchnorrPublicKey.fee(),
+        )
+        .await?;
+    generic::schnorr_public_key(arg).await
+}
+
+/// Signs a message using the caller's Schnorr key.
+///
+/// Note: This is an exact dual of the canister [`sign_with_schnorr`](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-sign_with_schnorr) method.  The argument and response types are also the same.
+///
+/// # Arguments
+/// - `arg`: The same `SignWithSchnorrArgument` as the management canister argument.  The semantics
+///   are identical but the meaning of the fields in the new context deserve some explanation.
+///   - `arg.canister_id`: The principal of the canister or user for which the Chain Fusion Signer
```

Comment:
Shouldn't this be the message instead of the canister_id?

Reply:
Good catch!

Reply:
Done.